include/rmm/cuda_stream.hpp

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/detail/error.hpp>
+
+#include <cuda_runtime_api.h>
+
+namespace rmm {
+
+/**
+ * @brief Owning wrapper for a CUDA stream.
+ *
+ * Provides RAII lifetime semantics for a CUDA stream.
+ *
+ */
+class cuda_stream {
+ public:
+  // No copy construction or assignment
+  cuda_stream(cuda_stream const&) = delete;
+  cuda_stream& operator=(cuda_stream const&) = delete;
+
+  // Move construction and assignment allowed
+  cuda_stream(cuda_stream&&) = default;
+  cuda_stream& operator=(cuda_stream&&) = default;
```

Comment:
This move logic will result in a double free, i.e., `cudaStreamDestroy` would be invoked in both the moved-to and moved-from object.

I suggest making the internal storage be a `unique_ptr<cudaStream_t>` as that way the copy/move functions can all be defaulted (rule of 0) and will have the correct behavior.

Reply:
Fixed, thanks.